### DISCUSSION TITLE
Using `typing.Any` as a return type changed behaviour after introducing response model declaration as return type in version `0.89.0`

### DISCUSSION THREAD

**Discussion by 15bubbles** [fastapi/fastapi#9626] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import datetime
from typing import Optional, Any

import uvicorn
from fastapi import APIRouter, Query, FastAPI
from pydantic import parse_obj_as
from pydantic.main import BaseModel

test_router = APIRouter(prefix="/test")

DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"


def to_camel(string: str) -> str:
    field_name = "".join(word.capitalize() for word in string.split("_"))
    return field_name[0].lower() + field_name[1:]


def format_date(dt: datetime) -> str:
    return dt.strftime(DATE_FORMAT)


class CamelModel(BaseModel):
    class Config:
        alias_generator = to_camel
        allow_population_by_field_name = True
        json_encoders = {
            datetime: format_date
        }


class Test(BaseModel):
    timestamp: datetime


class TestResponse(CamelModel):
    timestamp: datetime


@test_router.get("/items")
async def get_image_events(
        since: Optional[datetime] = Query(None),
) -> Any:
    test = Test(timestamp=since or datetime.utcnow())
    return parse_obj_as(TestResponse, test.dict())

app = FastAPI()
app.include_router(test_router)


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

Hi FastAPI team!

I'm seeking your guidance on this issue/question/matter (not sure how to call it to be honest).

Recently, we've updated FastAPI from `0.85.2` to `0.95.2`. Version `0.89.0` introduced a feature where you do not have to specify `response_model` but it get's inferred from the return type of route handler.

Now considering our simplified app implementation, that I've provided in "Example Code" section, before version `0.89.0`, our API returned such response:

```json
{"timestamp": "2023-06-06T11:16:32.359676Z"}
```

(notice `Z` in the end)

but since version `0.89.0` our API returns this response:

```json
{"timestamp": "2023-06-06T13:21:05.360710"}
```

(there's no `Z` at the end).

Now, putting aside the clear fact that the return type annotation is simply wrong, and we should put the `TestResponse` type there (this accident allowed us to find this problem) let's consider it in two cases:

## Before `0.89.0`

Now, we don't declare any `response_model` (which is possibly also wrong), and reading the response model from return type declaration feature is not avaiable yet. We have a return type of `Any` but that does not matter. This means, that with such setup the `response_model` [here](https://github.com/tiangolo/fastapi/blob/612b8ff168beee33fcf0c0bdff9718b4c8b1ac63/fastapi/routing.py#L326) will be `None`. Following this logic further, [this conditional](https://github.com/tiangolo/fastapi/blob/612b8ff168beee33fcf0c0bdff9718b4c8b1ac63/fastapi/routing.py#L392) will result in assigning `None` to the `self.response_field` as well. Skipping to `get_request_handler` and `serialize_response` function invocation, I presume that [this](https://github.com/tiangolo/fastapi/blob/612b8ff168beee33fcf0c0bdff9718b4c8b1ac63/fastapi/routing.py#L150.) is the condition branch that gets called.

This causes the response to get returned properly with such timestamp:

```json
{"timestamp": "2023-06-06T11:16:32.359676Z"}
```

and I presume that it uses `json_encoders` that are declared on the `BaseModel` extenstion that we use as our internal base model, because it uses the response that [gets returned](https://github.com/tiangolo/fastapi/blob/612b8ff168beee33fcf0c0bdff9718b4c8b1ac63/fastapi/routing.py#L235) from the endpoint (which is an instance of our model).

## Going to the `0.89.0`

We still have same setup, no `response_model` declared and `Any` as return type.,

Now, I think that the relevant change is in [this](https://github.com/tiangolo/fastapi/blob/0.89.0/fastapi/routing.py#L359) line. Having `Any` as a return type causes this to assign `Any` as a `self.response_model`. Because of that, [this conditional](https://github.com/tiangolo/fastapi/blob/0.89.0/fastapi/routing.py#L395), gets evaluated to `True` and `self.response_field` gets assigned a response field. Now, skipping to the [`serialize_response` function](https://github.com/tiangolo/fastapi/blob/69bd7d8501fe77601ebd67f55b39596fdd1e3792/fastapi/routing.py#L109) having a `field` that is truthy, it will obviously fall into the truthy branch which causes the response to get returned not in our desired way:

```json
{"timestamp": "2023-06-06T13:21:05.360710"}
```

Setting the return type to `None` (either explicitly or implicitly by not having any return type declared), solves this issue and the response that gets returned is with `Z` in the end.

## To the point

Now, after this long introduction and the fact that I know we've made an error setting `Any` or `None` (either implicitly or explicitly) as a return type (we simply overlooked it and we should type it better despite the introduction of recent FastAPI feature), let's get to the point:

I'm wondering if setting `Any` as a return type should work in such way and is it desired behaviour or just some side effect that was caused by the fact that it is "truthy"?

I mean, it kinda makes sense to do so, because it implies a bit, that we don't know what response model has to be used and we do it in a generic way. On the other hand it worked before `0.89.0` in such a way that when response model was not known explicitly, then FastAPI just used what got returned from the handler.

I hope that I've explained everything clearly.

I'd like to see what's your opinion on this topic! Thanks in advance!

### Operating System

Linux

### FastAPI Version

>=0.89.0

### Python Version

3.10

**Comment by Kludex** [comment]
Was this solved?
